### ISSUE TITLE
iTunes import broken after redesign

### ISSUE THREAD

**Issue by liftarn** [dufferzafar/Userscripts#8]
It appears the redesign broke the script.

**Comment by dufferzafar** [comment]
@liftarn @augustjanse I've updated the script and it now works for me. Please try and let me know!

**Comment by ewauq** [comment]
I can't see the importation button on any pages after the new design.

![image](http://i.imgur.com/fPcgQLl.jpg)

**Comment by liftarn** [comment]
Sorry, but I don't see the import button either.

**Comment by dufferzafar** [comment]
OK. This is weird since it works for me :confused: 

https://itunes.apple.com/fr/album/witness/id1236471560
![image](https://user-images.githubusercontent.com/1449512/27003831-d4458cde-4e1b-11e7-84ee-3c93e34a9b8d.png)

https://itunes.apple.com/us/album/best-of-kailash-kher/id420713987
![image](https://user-images.githubusercontent.com/1449512/27003838-f0c87182-4e1b-11e7-875d-06e3bd7fca31.png)

**Comment by dufferzafar** [comment]
How can we debug this now?

**Comment by ewauq** [comment]
Just deleted the userscript and re-installed it, same problem for me.

![image](http://i.imgur.com/CGfrh2M.jpg)

**Comment by dufferzafar** [comment]
Hi, considering that the "Link to HD Artwork" is working, I've moved the "Add to MusicBrainz" button to the left now. And have updated the script accordingly. See if this works.

![image](https://user-images.githubusercontent.com/1449512/27004795-8ca172a0-4e2d-11e7-98e9-5ce757c55b59.png)

**Comment by dufferzafar** [comment]
https://github.com/dufferzafar/Userscripts/commit/a14cbb0b6cfde86fde6d3e44695c6ff83b122dd6

**Comment by liftarn** [comment]
Odd. I first saw the Link to HD Artwork button, but then it vanished.
It may be an idea to bump the version as well.

**Comment by dufferzafar** [comment]
@liftarn When that happens, did you try and reload the page? It happened for me too once but I reloaded and it worked!

**Comment by V-Toll** [comment]
Hmm, it's not working for me either, even after the [latest update](https://github.com/dufferzafar/Userscripts/commit/a14cbb0b6cfde86fde6d3e44695c6ff83b122dd6). The link to HD artwork works (links only to the 939px version and not 1200px) but the import button is just not there (like liftarn and ewauq). I'm using Tampermonkey and Google Chrome 60.0.3112.24 an macOS 10.12.5.

**Comment by Bacardi89** [comment]
hey it don't work me.

**Comment by liftarn** [comment]
@dufferzafar Nope, just the artwork link first showing and then disappearing. Running Greasemonkey 3.11 on Firefox.

**Comment by liftarn** [comment]
I still haven't managed to get it to work using Firefox 54.0 and GreaseMonkey 3.11. I have tried to reinstall both the script and GreaseMonkey with no luck.

**Comment by augustjanse** [comment]
Did you check your console (`Ctrl+Shift+K`)? My update should have fixed the error visible in @ewauq's screenshot, if your console shows a different one that would be helpful. Also try refreshing the release page, seems like it can be a little slow in showing up.

**Comment by liftarn** [comment]
Noting really suspicious in the console:
`Cross-Origin begäran blockerad: Same-Origin policyn tillåter inte läsningar av fjärresurs på https://www.apple.com/ac/globalfooter/3/en_US/assets/ac-footer/legacy/appleicons_text.woff. (Orsak: CORS header 'Access-Control-Allow-Origin' saknas).  (okänt)
downloadable font: download failed (font-family: "Apple Legacy Chevron" style:normal weight:normal stretch:normal src index:1): bad URI or cross-site access not allowed source: https://www.apple.com/ac/globalfooter/3/en_US/assets/ac-footer/legacy/appleicons_text.woff  ac-globalfooter.built.css:1:1040
Metrics config: No config provided via delegate or fetched via init(), using default/cached config values. vendor-5f048e2ad0248e8169016946294565f8.js:28:11988
Cross-Origin begäran blockerad: Same-Origin policyn tillåter inte läsningar av fjärresurs på https://www.apple.com/ac/globalfooter/3/en_US/assets/ac-footer/legacy/appleicons_text.ttf. (Orsak: CORS header 'Access-Control-Allow-Origin' saknas).  (okänt)
downloadable font: download failed (font-family: "Apple Legacy Chevron" style:normal weight:normal stretch:normal src index:2): bad URI or cross-site access not allowed source: https://www.apple.com/ac/globalfooter/3/en_US/assets/ac-footer/legacy/appleicons_text.ttf  ac-globalfooter.built.css:1:1040
GET 
https://web-experience.itunes.apple.com/assets/controls/play@2x-b967ac20db4935540017bdc1ddbed546.svg [HTTP/1.1 200 OK 0ms]
OPTIONS 
XHR 
https://xp.apple.com/report/2/xp_its_preview [HTTP/1.1 200 OK 522ms]
POST 
XHR 
https://xp.apple.com/report/2/xp_its_preview [HTTP/1.1 200 OK 138ms]`

**Comment by liftarn** [comment]
When refreshing I sometimes get a "Link to HD artwork" that then disappear.

**Comment by augustjanse** [comment]
Replication succeeded! Windows 10, newly installed Firefox 54.0 64-bit and Greasemonkey 3.11 for the record, but the browser is probably the critical part. I'll take a look.

**Comment by liftarn** [comment]
Tested with your fix and it seem to solve the problem.

**Comment by libussa** [comment]
This is still broken for me. Link to HD artwork and import button appear briefly below the artwork, then disappears.
Firefox 55, Linux

**Comment by dufferzafar** [comment]
@libussa Link to itunes page where you're experiencing the problem?

A screenshot?

**Comment by libussa** [comment]
@dufferzafar everything I tried is problematic. A couple of examples: 
https://itunes.apple.com/us/album/good-will-continue-remixes/id1266082650
https://itunes.apple.com/fr/album/bubbling/id1205433832

screenshot: https://i.imgur.com/eZKvKkr.png

**Comment by liftarn** [comment]
Yes, it appears and then disappears. Using Firefox 54.0.1 and Greasemonkey 3.11

**Comment by dufferzafar** [comment]
Are you sure you guys are using the right version (current master tip)

Because it stays visible for me. I'm on KDE Neon using FF 55.0.2 and the [dotjs addon](https://github.com/rlr/dotjs-addon).

**Comment by liftarn** [comment]
Yes, reinstalled from https://github.com/dufferzafar/Userscripts/blob/master/MB-Import-From-iTunes.user.js today and tested.

**Comment by Bacardi89** [comment]
Same don't work

Obtenez Outlook pour iOS<https://aka.ms/o0ukef>
________________________________

**Comment by otringal** [comment]
I had the same problem but found a solution that works for me. I think the problem is that the script runs before the page has finished loading. Add this at the top of the script if you use Greasemonkey:
`//@run-at  document-idle`

**Comment by AzoresOne** [comment]
@otringal 

> Add this at the top of the script if you use Greasemonkey:
//@run-at document-idle

I get this warning:
**Scam Warning: Take care when pasting things you don't understand. This could allow attackers to steal
your identity or take control of your computer. Please type 'allow pasting' in the scratchpad below to allow pasting**.

Now what?

**Comment by AzoresOne** [comment]
Update:
I went ahead and typed instead of pasting and whattya know, it now works properly!
Thanks 👍

**Comment by V-Toll** [comment]
Strange. Same problem here, even with `//@run-at document-idle`

Here is an example: the US iTunes site doesn't work:
https://itunes.apple.com/us/album/raining-single/id1272276768
<img width="1296" alt="01" src="https://user-images.githubusercontent.com/7504584/29739670-ca999432-8a43-11e7-97e7-268dfcc59c8d.png">

The German one works fine for me:
https://itunes.apple.com/de/album/raining-single/id1272276768
<img width="1253" alt="2" src="https://user-images.githubusercontent.com/7504584/29739671-d8756ff4-8a43-11e7-8ae6-48e500bbea61.png">

**Comment by liftarn** [comment]
Even with //@run-at document-idle it don't always appears so I often have to reload the page to make it appear, but it's still a vast improvement over how it was before.

**Comment by augustjanse** [comment]
I haven't had this problem recently, but started encountering it today. These two were the problem at first:

- https://itunes.apple.com/se/album/s%C3%A5-mycket-b%C3%A4ttre-tolkningarna-s%C3%A4song-7/id1202299483
- https://itunes.apple.com/se/album/tolkningarna-s%C3%A5-mycket-b%C3%A4ttre-s%C3%A4song-7/id1183961524

But now that I test other albums, they all seem to be having problems, including ones I had no problem with just now. No idea what's the problem. Don't really have any time to debug it either, sorry.

Tried @Mirgolth's fix, but it didn't do much. It seems reasonable, but on the other hand the script already [waits for a state change](https://github.com/dufferzafar/Userscripts/blob/master/MB-Import-From-iTunes.user.js#L23) so I'm not sure it does much.

**Comment by augustjanse** [comment]
I found while debugging that pausing the execution of the userscript (for example by setting debugging mode in Tampermonkey/Greasemonkey and opening the console before updating the page) consistently makes the button appear as it should after resuming. It seems clear that the problem is when it executes before the iTunes scripts do. Not sure how to fix it, but this is a pretty useful workaround.

**Comment by jesus2099** [comment]
Hello, I also have a broken iTunes script I don’t really intend to work on any more.

But just to say I have noticed that iTunes JavaScript is running routinely to red raw the page (I see an ugly font jump, few seconds after each page load in opera presto).

So I have put a delay before my script runs and it worked but it is not good solution.
https://github.com/jesus2099/konami-command/issues/353#issuecomment-326946527

**Comment by 5moufl** [comment]
works again on Firefox 58 with ViolentMonkey

**Comment by gauravjuvekar** [comment]
Disabling `security.csp.enabled` in FF ESR 52.7.3 allows it to work correctly.

**Comment by 5moufl** [comment]
seems to be broken again in Firefox 59.0.3 :(
works fine in Chrome 66.0.3359.117
I use with ViolentMonkey 2.9.4 with both.

**Comment by jesus2099** [comment]
@dufferzafar, you could try to add a `setInterval(function() { … }, 5000);` that keeps checking if your buttons are still there, and if not, readds them.

**Comment by jesus2099** [comment]
@Mirgolth’s fix #15 (`// @run-at document-idle`), which is a sensible improvement in running this script on iTunes’ new page design, has just been deployed!

**Comment by nitroeskimo** [comment]
I cannot see the "Add to MusicBrainz" button with Firefox 65.0.2 and GreaseMonkey 4.7 so I install ViolentMonkey 2.10.3 and it works !

**Comment by jesus2099** [comment]
Does anyone still have this problem since version `2017.11.21.1` deployed on **2019-02-08**?
**No problem here with [Vivaldi](https://vivaldi.com) (Chrome) + [Violentmonkey](https://chrome.google.com/webstore/detail/violentmonkey/jinjaccalgkegednnccohejagnlnfdag).**

**Comment by jesus2099** [comment]
I forgot to mention the very easy work around: Press F5 to load page properly and to let the user script catch up.
